### PR TITLE
RB - Added ability to edit tags

### LINKED ISSUE #14: Modify Tag List
#### As a user I want to edit the tag list for an image.

__Story__:

Image tags are awesome -- our user's love them! Tags help us make sense of the
plethora of images added to our system. But how annoying is it when our users
make spelling mistakes in their image tags? Super annoying! How awesome, and
useful would it be to allow for tags to be edited after an image is added?
Super awesome, and useful!

__Acceptance criteria__:
- [ ] There is a link to edit the tags for an image.
- [ ] The image whose tags are to be modified is displayed on the edit page.
- [ ] On the edit page, I can change the tags to any valid tag list.
- [ ] After saving the edit page with valid tags, my changes are visible
  throughout the rest of the application.

__Discussion__:
- [ ] Should an image's URL be editable?

__Dependencies__:
- Image Tags
- Refactor flow tests to use page objects

### PR DESCRIPTION
**Pull request by RitamBanerjee**
Closes #14 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [ ] I have replaced `X` above with the associated issue number.
- [ ] I have changed the issue's label to "pull request".
- [ ] I have assigned the designated reviewer to this PR.
- [ ] I have set the appropriate label on this PR.
- [ ] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: app/controllers/images_controller.rb [hunk 1]
```diff
@@ -34,6 +34,18 @@ def index
               end
   end
 
+  def edit
+  end
+
+  def update
+    if @image.update_attributes(permitted_update_image_params)
+      flash[:success] = 'Tags Updated'
+      redirect_to image_path(@image.id)
```
Comment (wilsonyqm) [inline app/controllers/images_controller.rb]: no need for id here

File: app/views/images/show.html.erb [hunk 2]
```diff
@@ -10,6 +10,8 @@
 </p>
 
 <ul>
+  <%= link_to 'Edit', edit_image_path(@image.id),
```
Comment (wilsonyqm) [inline app/views/images/show.html.erb]: can you make the line below to be the same line with the current line, and one more whitespace before the `%>`

File: config/routes.rb [hunk 3]
```diff
@@ -5,7 +5,7 @@
 
   resources :users, only: [:create, :new]
 
-  resources :images, only: [:create, :show, :new, :index, :destroy] do
+  resources :images, only: [:create, :show, :new, :index, :destroy, :edit, :update] do
```
Comment (wilsonyqm) [inline config/routes.rb]: no more `only` is needed

File: test/flows/images_flow_test.rb [hunk 4]
```diff
@@ -60,6 +60,18 @@ class ImagesFlowTest < FlowTestCase
     assert images_index_page.showing_image?(url: cute_puppy_url)
   end
 
+  test 'update an image' do
+    cute_puppy_url = 'https://i.imgur.com/qOWKUb.jpg'
+    cute_puppy_tags = %w(puppy cute)
+    new_cute_puppy_tags = %w(puppy scary)
+    Image.create!([{ title: 'dog', url: cute_puppy_url, tag_list: cute_puppy_tags.join(', ') }])
```
Comment (wilsonyqm) [inline test/flows/images_flow_test.rb]: can you test the error case in selenium test?

File: test/controllers/images_controller_test.rb [hunk 5]
```diff
@@ -120,4 +120,14 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
...
   end
+
+  test 'Tags can be updated' do
+    @image = Image.create!(title: 'Test Image4', url: 'http://i.imgur.com/qkdpN.jpg', tag_list: 'tag4, tag3')
+    tag_list_params = { tag_list: 'tag1, tag2, tag3' }
+
+    patch image_path(@image), params: { image: tag_list_params }
+    assert_redirected_to image_path(@image.id)
+    @image.reload
+    assert_equal @image.tag_list.count, 3
```
Comment (wilsonyqm) [inline test/controllers/images_controller_test.rb]: can you test some error cases in your controller test?

**Comment by wilsonyqm** [comment]
In the description of the story, Can you mark which issue it refers? Closes #??

**Comment by wilsonyqm** [comment]
Can you think of any corner cases which would make the tag_lists fail? or the UI or the acts as taggable is not well handled?

File: test/controllers/images_controller_test.rb [hunk 6]
```diff
@@ -137,4 +137,24 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
...
+  end
+
+  test 'Tags can not be set to empty' do
+    @image = Image.create!(title: 'Test Image4', url: 'http://i.imgur.com/qkdpN.jpg', tag_list: 'tag4, tag3')
+    tag_list_params = { tag_list: '' }
+
+    patch image_path(@image), params: { image: tag_list_params }
+    assert_select '.image_tag_list > .help-block', text: 'can\'t be blank'
```
Comment (wilsonyqm) [inline test/controllers/images_controller_test.rb]: you don't need the `>` I think?

File: test/controllers/images_controller_test.rb [hunk 7]
```diff
@@ -137,4 +137,24 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
...
+    @image.reload
+    assert_equal @image.tag_list.count, 2
+  end
```
Comment (wilsonyqm) [inline test/controllers/images_controller_test.rb]: How did you handle the very long tag case? can you test it too?

File: app/views/images/edit.html.erb [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+      <%= f.input :tag_list,label: 'New Tag List', input_html: { value: @image.tag_list.join(', ').to_s } %>
+      <%= f.button :submit, class: 'btn btn-success'%>
+  <% end %>
+</div>
+
+<p>
+  <img class='js-show-image' src='<%=@image.url%>'>
+</p>
+<p>
+  <%= render partial: 'images/image_tags', locals: {image: @image} %>
```
Comment (wilsonyqm) [inline app/views/images/edit.html.erb]: remove this line plz

File: app/models/image.rb [hunk 9]
```diff
@@ -3,4 +3,11 @@ class Image < ApplicationRecord
   validates :url, presence: true
   validates :url, format: { with: %r{\Ahttps?:\/\/[a-zA-Z0-9\.\/]+\.(gif|jpe?g|png|bmp|tiff)\z} }
   validates :tag_list, presence: true
+  validate :max_tag_size
+
+  def max_tag_size
+    longest_tag = 0
```
Comment (wilsonyqm) [inline app/models/image.rb]: This suppose to be a tag, not the length, I think

File: app/models/image.rb [hunk 10]
```diff
@@ -3,4 +3,11 @@ class Image < ApplicationRecord
...
+    tag_list.each { |tag| longest_tag = tag if tag.length > longest_tag.to_i }
```
Comment (wilsonyqm) [inline app/models/image.rb]: tag.to_i doesn't make sense to me.

File: test/controllers/images_controller_test.rb (hunk 6 above)
Comment (wilsonyqm) [reply test/controllers/images_controller_test.rb]: Have you tried to remove the `>` sign?

File: test/models/image_test.rb [hunk 11]
```diff
@@ -83,4 +83,10 @@ def test_image_tags_saved
     assert_predicate @image, :valid?
     assert_equal %w(tag1 tag2 tag3), @image.tag_list
   end
+
+  def tags_invalid_if_too_long
+    @image = Image.new(title: 'Doggo1', url: 'http://i.imgur.com/J23eb35.bmp', tag_list: 'tags_longer_than_20_characters_are_unnecessary_')
+
+    refute_predicate @image, :valid?
```
Comment (wilsonyqm) [inline test/models/image_test.rb]: What's more you need to assert?

File: test/controllers/images_controller_test.rb (hunk 7 above)
Comment (wilsonyqm) [reply test/controllers/images_controller_test.rb]: Can you test it here(not only in image model) as well?

File: app/models/image.rb [hunk 12]
```diff
@@ -3,4 +3,11 @@ class Image < ApplicationRecord
   validates :url, presence: true
   validates :url, format: { with: %r{\Ahttps?:\/\/[a-zA-Z0-9\.\/]+\.(gif|jpe?g|png|bmp|tiff)\z} }
   validates :tag_list, presence: true
+  validate :max_tag_size
+
+  def max_tag_size
+    longest_tag_length = 0
+    tag_list.each { |tag| longest_tag_length = tag.length if tag.length > longest_tag_length }
```
Comment (wilsonyqm) [inline app/models/image.rb]: One another cool way is to use `any?` in ruby, can you try to do it?

**Comment by wilsonyqm** [comment]
Looks good to me, you did a great job!!! Approved!
